### PR TITLE
Fix snippet overtyping feature

### PR DESCRIPTION
**Pull request by n-gist**
Fix snippet overtyping feature #105440 which stopped working during 1.72 ... 1.73 due to 1e662b070fd9f847dfefbdf3f45d0fe538d7e974

### CODE REVIEW

**Comment by n-gist** [comment]
@jrieken, I just moved model cancellation call a line later because, if I'm right, the feature gets the selected text from the model, so it must exist at the time of insertion but perhaps a better solution is needed

File: src/vs/editor/contrib/suggest/browser/suggestController.ts [hunk 1]
```diff
@@ -393,6 +390,9 @@ export class SuggestController implements IEditorContribution {
 			overtypingCapturer: this._overtypingCapturer.value
 		});
 
+		// cancel -> stops all listening and closes widget
+		this.model.cancel();
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/browser/suggestController.ts:394]: The cancelling must happen before calling `insert`. It needs a different change which freezes the overtyping capturer before that.

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/suggest/browser/suggestController.ts (hunk 1 above)
Comment (n-gist) [reply src/vs/editor/contrib/suggest/browser/suggestController.ts]: Reverted.
Changed logic of emptying captured selections and added locking while suggestions widget is active.
This should also fix some non-critical cases when it returned previously saved selection when there were no selection at widget trigger time.

**Review by jrieken — APPROVED** [review]
lgtm

**Comment by jrieken** [comment]
Thanks @n-gist

**Review by aeschli — APPROVED** [review]

**Comment by Walker-Xin** [comment]
The issue came back after 17.4.2 update.

**Comment by n-gist** [comment]
> The issue came back after 17.4.2 update.

1.74.2 working as expected for me

**Pull request merged**